src/Microsoft.AspNetCore.Routing/EndpointNameEndpointFinder.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    internal class EndpointNameEndpointFinder : IEndpointFinder<string>
+    {
+        private readonly DataSourceDependentCache<Dictionary<string, Endpoint[]>> _cache;
+
+        public EndpointNameEndpointFinder(CompositeEndpointDataSource dataSource)
+        {
+            _cache = new DataSourceDependentCache<Dictionary<string, Endpoint[]>>(dataSource, Initialize);
+        }
+
+        // Internal for tests
+        internal Dictionary<string, Endpoint[]> Entries => _cache.EnsureInitialized();
+
+        public IEnumerable<Endpoint> FindEndpoints(string address)
+        {
+            if (address == null)
+            {
+                throw new ArgumentNullException(nameof(address));
+            }
+
+            var entries = Entries;
+            entries.TryGetValue(address, out var result);
+            return result ?? Array.Empty<Endpoint>();
+        }
+
+        private static Dictionary<string, Endpoint[]> Initialize(IReadOnlyList<Endpoint> endpoints)
+        {
+            // Collect duplicates as we go, blow up on startup if we find any.
+            var duplicates = new Dictionary<string, List<Endpoint>>(StringComparer.Ordinal);
```

Comment:
I did a different change in the same spirit here. Rather than lazily building a list, I just use the `entries` to check what names I've seen. If I see a duplicate, I bail and do a linq.